libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineView.kt

```diff
@@ -154,11 +157,46 @@ class MapboxRouteLineView(options: MapboxRouteLineOptions) {
         MASKING_LAYER_RESTRICTED
     )
 
+    private val routeRenderedObservers = CopyOnWriteArraySet<RouteRenderedObserver>()
+    private val onSourceDataLoadedListener = OnSourceDataLoadedListener { eventData ->
+        if (eventData.loaded == true) {
+            val routeId = sourceToFeatureMap[RouteLineSourceKey(eventData.id)]?.routeId
+            routeId?.let {
+                routeRenderedObservers.forEach { observer -> observer.onRouteRendered(it) }
+            }
+        }
+    }
+
     @TestOnly
     internal fun initPrimaryRouteLineLayerGroup(layerIds: Set<String>) {
         primaryRouteLineLayerGroup = layerIds
     }
 
+    /**
+     * Register [RouteRenderedObserver].
+     * NOTE: the observer is not sticky, i. e. you will NOT be notified of previously rendered routes.
+     *
+     * @param map [MapboxMap] which the route is expected to be rendered on
+     * @param observer observer to add
+     */
+    fun registerRouteRenderedObserver(map: MapboxMap, observer: RouteRenderedObserver) {
```

Comment:
I don't think this is a good idea. This class was intended from the beginning NOT to hold a reference to the Map or the Style.  We've always told customers that this class doesn't hold any references to the Style and therefore the class could be created and destroyed at will.  In addition there are customers who create multiple instances of this class. Introducing this observer here potentially creates unforeseen issues for customers going forward.  

From what I can see all that is needed is the routeID when the observer event fires.  The routeID is available in `Feature::id` of the `FeatureCollection` in `RouteLineData`.  The functionality needed can be abstracted into a separate lightweight class that takes the same `Expected<RouteLineError, RouteSetValue>>` as the `renderRouteDrawData` in `MapboxRouteLineView`. That would keep the map observer out of the class.  

Also, I"m not sure what is meant by "rendered" in the original issue.  Drawing the `FeatureCollection` is one step but another step is updating the gradients.  Is the approach of notifying only the map source update sufficient?

Reply:
> Is the approach of notifying only the map source update sufficient?

The main problem is that the incidents are drawn separately and sometimes they are drawn before the route line. For that purpose I think drawing the `FeatureCollection` is enough.

I'll follow-up with the rest of the comment separately.